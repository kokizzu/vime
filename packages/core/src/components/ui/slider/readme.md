# vime-slider

A custom styled and ARIA friendly `input[type="range"]` component for inputting numeric values.
In addition, there are optimizations made for improved touch support (more information can be found
at https://github.com/sampotts/rangetouch).

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/slider/slider.png"
  alt="Vime slider component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html title="example.html"
<vime-slider
  step="5"
  max="100"
  label="Volume"
  [value]="value"
  (vValueChange)="onValueChange($event)"
/>
```

```ts title="example.ts"
class Example {
  value = 50;

  onValueChange(event: CustomEvent<number>) {
    this.value = event.detail;
  }
}
```

### Html

```html
<vime-slider step="5" max="100" value="50" label="Volume" />

<script>
  const slider = document.querySelector('vime-slider');

  slider.addEventListener('vValueChange', (event) => {
    const newValue = event.detail;
  });
</script>
```

### React

```tsx {2,12-18}
import React, { useState } from 'react';
import { VimeSlider } from '@vime/react';

function Example() {
  const [value, setValue] = useState(50);

  const onValueChange = (event: CustomEvent<number>) => {
    setValue(event.detail);
  };

  return render(
    <VimeSlider
      step="5"
      max="100"
      label="Volume"
      value={value}
      onVValueChange={onValueChange}
    />
  );
}
```

### Vue

```html {2-8,12,16} title="example.vue"
<template>
  <VimeSlider
    step="5"
    max="100"
    label="Volume"
    :value="value"
    @vValueChange="onValueChange"
  />
</template>

<script>
  import { VimeSlider } from '@vime/vue';

  export default {
    components: {
      VimeSlider,
    },

    data: {
      value: 50,
    },

    methods: {
      onValueChange(value) {
        this.value = value;
      },
    },
  };
</script>
```

## Properties

| Property    | Attribute    | Description                                                                                | Type                  | Default     |
| ----------- | ------------ | ------------------------------------------------------------------------------------------ | --------------------- | ----------- |
| `label`     | `label`      | A human-readable label for the purpose of the slider.                                      | `string \| undefined` | `undefined` |
| `max`       | `max`        | The greatest permitted value.                                                              | `number`              | `10`        |
| `min`       | `min`        | The lowest value in the range of permitted values.                                         | `number`              | `0`         |
| `step`      | `step`       | A number that specifies the granularity that the value must adhere to.                     | `number`              | `1`         |
| `value`     | `value`      | The current value.                                                                         | `number`              | `5`         |
| `valueText` | `value-text` | Human-readable text alternative for the current value. Defaults to `value:max` percentage. | `string \| undefined` | `undefined` |

## Events

| Event          | Description                                                     | Type                  |
| -------------- | --------------------------------------------------------------- | --------------------- |
| `vValueChange` | Emitted when the value of the underlying `input` field changes. | `CustomEvent<number>` |

## CSS Custom Properties

| Name                            | Description                                                       |
| ------------------------------- | ----------------------------------------------------------------- |
| `--slider-thumb-bg`             | The background color of the slider thumb.                         |
| `--slider-thumb-height`         | The height of the slider thumb.                                   |
| `--slider-thumb-shadow`         | The shadow cast around the slider thumb.                          |
| `--slider-thumb-width`          | The width of the slider thumb.                                    |
| `--slider-track-color`          | The color of the track.                                           |
| `--slider-track-focused-height` | The height of the track when it is focused.                       |
| `--slider-track-height`         | The height of the track.                                          |
| `--slider-value-color`          | The color of the part of the track filled upto the current value. |

## Dependencies

### Used by

- [vime-scrubber-control](../controls/scrubber-control)
- [vime-volume-control](../controls/volume-control)

### Graph

```mermaid
graph TD;
  vime-scrubber-control --> vime-slider
  vime-volume-control --> vime-slider
  style vime-slider fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
