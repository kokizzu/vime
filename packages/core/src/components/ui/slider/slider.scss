vime-slider {
  /**
  * @prop --slider-thumb-bg: The background color of the slider thumb.
  * @prop --slider-thumb-width: The width of the slider thumb.
  * @prop --slider-thumb-height: The height of the slider thumb.
  * @prop --slider-thumb-shadow: The shadow cast around the slider thumb.
  * @prop --slider-track-height: The height of the track.
  * @prop --slider-track-focused-height: The height of the track when it is focused.
  * @prop --slider-track-color: The color of the track.
  * @prop --slider-value-color: The color of the part of the track filled upto the current value.
  */

  $slider-thumb-width: var(--slider-thumb-width);
  $slider-thumb-height: var(--slider-thumb-height);
  $slider-thumb-bg: var(--slider-thumb-bg);
  $slider-thumb-border: 2px solid transparent;
  $slider-thumb-shadow: var(--slider-thumb-shadow);
  
  $slider-track-height: var(--slider-track-height);
  $slider-track-focused-height: var(--slider-track-focused-height);

  $slider-track-color: var(--slider-track-color);
  $slider-value-color: var(--slider-value-color);

  @mixin slider-track() {
    background: transparent;
    border: 0;
    border-radius: calc(#{$slider-track-height} / 2);
    height: $slider-track-height;
    transition: box-shadow 0.3s ease;
    user-select: none;
  }

  @mixin slider-thumb() {
    opacity: 0;
    background: $slider-thumb-bg;
    border: 0;
    border-radius: 100%;
    position: relative;
    transition: all 0.2s ease;
    width: $slider-thumb-width;
    height: $slider-thumb-height;
    box-shadow: $slider-thumb-shadow;
  }

  input {
    -webkit-appearance: none;
    background: transparent;
    border: 0;
    outline: 0;
    cursor: pointer;
    border-radius: calc(#{$slider-thumb-height} * 2);
    user-select: none;
    -webkit-user-select: none;
    touch-action: manipulation;

    // Color is used in JS to populate lower fill for WebKit
    color: $slider-value-color;
    display: block;
    height: $slider-track-height;
    margin: 0;
    padding: 0;
    transition: box-shadow 0.3s ease;
    width: 100%;

    &::-webkit-slider-runnable-track {
      @include slider-track();

      background-image: linear-gradient(to right, currentColor var(--value, 0%), transparent var(--value, 0%));
      background-color: $slider-track-color;
    }

    &::-webkit-slider-thumb {
      @include slider-thumb();
      
      -webkit-appearance: none;
      margin-top: calc(0px - calc(calc(#{$slider-thumb-height} - #{$slider-track-height}) / 2));
    }

    &::-moz-range-track {
      @include slider-track();
      background-color: $slider-track-color;
    }

    &::-moz-range-thumb {
      @include slider-thumb();
    }

    &::-moz-range-progress {
      background: currentColor;
      border-radius: calc(#{$slider-track-height} / 2);
      height: $slider-track-height;
    }

    &::-ms-track {
      @include slider-track();

      color: transparent;
      background-color: $slider-track-color;
    }

    &::-ms-fill-upper {
      @include slider-track();
    }

    &::-ms-fill-lower {
      @include slider-track();

      background: currentColor;
    }

    &::-ms-thumb {
      @include slider-thumb();

      // For some reason, Edge uses the -webkit margin above
      margin-top: 0;
    }

    &::-ms-tooltip {
      display: none;
    }

    &:hover {
      &::-webkit-slider-runnable-track {
        height: $slider-track-focused-height;
      }

      &::-moz-range-track {
        height: $slider-track-focused-height;
      }

      &::-ms-track {
        height: $slider-track-focused-height;
      }

      &::-ms-fill-upper {
        height: $slider-track-focused-height;
      }

      &::-ms-fill-lower {
        height: $slider-track-focused-height;
      }

      &::-webkit-slider-thumb {
        opacity: 1;
      }

      &::-moz-range-thumb {
        opacity: 1;
      }

      &::-ms-thumb {
        opacity: 1;
      }
    }

    &:focus {
      outline: 0;

      &::-webkit-slider-runnable-track {
        outline: 0;
        height: $slider-track-focused-height;
      }

      &::-moz-range-track {
        outline: 0;
        height: $slider-track-focused-height;
      }

      &::-ms-track {
        outline: 0;
        height: $slider-track-focused-height;
      }
    }

    &::-moz-focus-outer {
      border: 0;
    }
  }
}