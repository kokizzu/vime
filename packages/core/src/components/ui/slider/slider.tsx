import {
  h, Host, Component, Prop, Event, EventEmitter, Element,
} from '@stencil/core';

@Component({
  tag: 'vime-slider',
  styleUrl: 'slider.scss',
})
export class Slider {
  @Element() el!: HTMLVimeSliderElement;

  /**
   * A number that specifies the granularity that the value must adhere to.
   */
  @Prop() step = 1;

  /**
   * The lowest value in the range of permitted values.
   */
  @Prop() min = 0;

  /**
   * The greatest permitted value.
   */
  @Prop() max = 10;

  /**
   * The current value.
   */
  @Prop() value = 5;

  /**
   * Human-readable text alternative for the current value. Defaults to `value:max` percentage.
   */
  @Prop() valueText?: string;

  /**
   * A human-readable label for the purpose of the slider.
   */
  @Prop() label?: string;

  /**
   * Emitted when the value of the underlying `input` field changes.
   */
  @Event() valueChange!: EventEmitter<number>;

  private getPercentage() {
    return `${(this.value / this.max) * 100}%`;
  }

  private onValueChange(event: Event) {
    const value = parseFloat((event.target as HTMLInputElement)?.value);
    this.valueChange.emit(value);
  }

  private calcTouchedValue(event: TouchEvent) {
    const input = event.target! as HTMLInputElement;
    const touch = event.changedTouches[0];
    const min = parseFloat(input.getAttribute('min')!);
    const max = parseFloat(input.getAttribute('max')!);
    const step = parseFloat(input.getAttribute('step')!);
    const delta = max - min;

    // Calculate percentage.
    let percent;
    const clientRect = input.getBoundingClientRect();

    const sliderThumbWidth = parseFloat(
      window.getComputedStyle(this.el).getPropertyValue('--slider-thumb-width'),
    );

    const thumbWidth = ((100 / clientRect.width) * (sliderThumbWidth / 2)) / 100;

    percent = (100 / clientRect.width) * (touch.clientX - clientRect.left);

    // Don't allow outside bounds.
    percent = Math.max(0, Math.min(percent, 100));

    // Factor in the thumb offset.
    if (percent < 50) {
      percent -= (100 - percent * 2) * thumbWidth;
    } else if (percent > 50) {
      percent += (percent - 50) * 2 * thumbWidth;
    }

    const position = delta * (percent / 100);

    if (step >= 1) {
      return min + Math.round(position / step) * step;
    }

    /**
     * This part differs from original implementation to save space. Only supports 2 decimal
     * places (0.01) as the step.
     */
    return min + parseFloat(position.toFixed(2));
  }

  /**
   * Basically input[range="type"] on touch devices sucks (particularly iOS), so this helps make it
   * better.
   *
   * @see https://github.com/sampotts/rangetouch
   */
  private onTouch(event: TouchEvent) {
    const input = event.target! as HTMLInputElement;
    if (input.disabled) return;
    event.preventDefault();
    this.value = this.calcTouchedValue(event);
    this.valueChange.emit(this.value);
    input.dispatchEvent(
      new window.Event((event.type === 'touchend') ? 'change' : 'input', { bubbles: true }),
    );
  }

  render() {
    return (
      <Host
        style={{
          '--value': this.getPercentage(),
        }}
      >
        <input
          type="range"
          step={this.step}
          min={this.min}
          max={this.max}
          value={this.value}
          autocomplete="off"
          aria-label={this.label}
          aria-valuemin={this.min}
          aria-valuemax={this.max}
          aria-valuenow={this.value}
          aria-valuetext={this.valueText ?? this.getPercentage()}
          aria-orientation="horizontal"
          onInput={this.onValueChange.bind(this)}
          onFocus={() => { this.el.dispatchEvent(new window.Event('focus', { bubbles: true })); }}
          onBlur={() => { this.el.dispatchEvent(new window.Event('blur', { bubbles: true })); }}
          onTouchStart={this.onTouch.bind(this)}
          onTouchMove={this.onTouch.bind(this)}
          onTouchEnd={this.onTouch.bind(this)}
        />
      </Host>
    );
  }
}
