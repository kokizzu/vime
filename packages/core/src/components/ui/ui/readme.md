# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
