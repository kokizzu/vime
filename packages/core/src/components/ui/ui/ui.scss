vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}