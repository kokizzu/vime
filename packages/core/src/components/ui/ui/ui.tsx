import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.Playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.IsFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        {canShowCustomUI && <slot />}
      </Host>
    );
  }
}

openPlayerWormhole(UI, [
  PlayerProp.IsVideoView,
  PlayerProp.Playsinline,
  PlayerProp.IsFullscreenActive,
]);
