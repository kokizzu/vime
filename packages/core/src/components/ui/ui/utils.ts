import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
