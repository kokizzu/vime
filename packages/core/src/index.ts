import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
