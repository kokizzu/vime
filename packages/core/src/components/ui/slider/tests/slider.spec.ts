import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../ui/tests';
import { Slider } from '../slider';

let page: SpecPage;
let slider: HTMLVimeSliderElement;
let input: HTMLInputElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [Slider],
    '<vime-slider />',
  ));

  slider = page.root!.querySelector('vime-slider')!;
  input = page.root!.querySelector('input')!;
});

it('should be structurally sound', () => {
  expect(slider).toMatchSnapshot();
});

it('should set the value text', async () => {
  expect(input.getAttribute('aria-valuetext')).toEqual('50%');
  slider.value = 7;
  await page.waitForChanges();
  expect(input.getAttribute('aria-valuetext')).toEqual('70%');
});

it('should fire valueChange event', async () => {
  const callback = jest.fn();
  slider.addEventListener('valueChange', callback);
  input.value = '8';
  input.dispatchEvent(new Event('input'));
  await page.waitForChanges();
  expect(callback).toHaveBeenCalled();
});

it('should fire focus event', async () => {
  const callback = jest.fn();
  slider.addEventListener('focus', callback);
  input.dispatchEvent(new Event('focus'));
  await page.waitForChanges();
  expect(callback).toHaveBeenCalled();
});

it('should fire blur event', async () => {
  const callback = jest.fn();
  slider.addEventListener('blur', callback);
  input.dispatchEvent(new Event('blur'));
  await page.waitForChanges();
  expect(callback).toHaveBeenCalled();
});
