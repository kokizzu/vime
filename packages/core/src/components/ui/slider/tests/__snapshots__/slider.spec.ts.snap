// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-slider style="--value: 50%;">
  <input aria-orientation="horizontal" aria-valuemax="10" aria-valuemin="0" aria-valuenow="5" aria-valuetext="50%" autocomplete="off" max="10" min="0" step="1" type="range" value="5">
</vime-slider>
`;
